python/osrd_schemas/scripts/truncate_infra.py

```diff
@@ -0,0 +1,378 @@
+                print(f"Warning: dropping {route.id} that is partially included")
+
+        elif entry_included or exit_included:
+            # The route is partially included, we need to adapt it
+            route = route.model_copy()
+            switch_directions = order_switch_directions(
+                route, detectors_by_id, buffer_stops_by_id, switches_by_id
+            )
+            if not exit_included:
+                # It means the exit point is not in the bbox so we look from the end
+                switch_directions = list(reversed(switch_directions))
+
+            idx = 0
+            while idx < len(switch_directions) and (
+                switch_directions[idx][0] not in switch_ids_kept
+            ):
+                idx += 1
+            assert all(
+                switch_id in switch_ids_kept for switch_id, _ in switch_directions[idx:]
+            )
+
+            new_bufferstop_id, direction = former_switch_id_to_buffer_stop_id[
+                switch_directions[idx - 1][0]
+            ]
+            if entry_included:
+                route.exit_point = BufferStopReference(id=new_bufferstop_id)
+            else:
+                route.entry_point = BufferStopReference(id=new_bufferstop_id)
+                route.entry_point_direction = direction
+
+            route.switches_directions = dict(switch_directions[idx:])
+            route.release_detectors = [
+                d for d in route.release_detectors if d in waypoints_ids_kept
+            ]
+
+            res.append(route)
+
+        elif not entry_included and not exit_included and n_switches_included > 0:
+            # Some edge case that should be included but isn't handled yet
+            print(f"Weird route {route.id} that should be included but isn't handled")
+    return res
+
+
+def order_switch_directions(
+    route: Route,
+    detectors_by_id: dict[str, Detector],
+    buffer_stops_by_id: dict[str, BufferStop],
+    switches_by_id: dict[str, Switch],
+) -> list[tuple[str, str]]:
+    """Order the switches and their directions in the route from the entry point to the exit point."""
+    res = []
+
+    tracks_by_switch = {
+        switch_id: [port.track for port in switch.ports.values()]
+        for switch_id, switch in switches_by_id.items()
+    }
+
+    switch_directions = list(route.switches_directions.items())
+    entry_waypoint = (
+        detectors_by_id[route.entry_point.id]
+        if route.entry_point.type == "Detector"
+        else buffer_stops_by_id[route.entry_point.id]
+    )
+    track_id = entry_waypoint.track
+    while switch_directions != []:
+        switch_dir_on_track = [
+            (switch_id, dir)
+            for switch_id, dir in switch_directions
+            if track_id in tracks_by_switch[switch_id]
+        ]
+
+        assert len(switch_dir_on_track) == 1
+        res.append(switch_dir_on_track[0])
+        switch_id, dir = res[-1]
+
+        switch_directions.remove(res[-1])
```

Comment:
nitpick: Isn’t it more straightforward to understand when we use directly the object instead of the indirection of “the last element of an array in which we just inserted an element”?


```suggestion
        switch_id, dir = switch_dir_on_track[0]

        switch_directions.remove(switch_dir_on_track[0])
```

Reply:
Done in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/e2bfeae553e0f297c3b037aa9d351fbd67b3fe2d.